hack/compare-k8s-feature-gates.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+#
+# Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+usage() {
+  echo "Usage:"
+  echo "> compare-k8s-feature-gates.sh [ -h | <old version> <new version> ]"
+
+  exit 0
+}
+
+if [ "$1" == "-h" ] || [ "$#" -ne 2 ]; then
+  usage
+fi
+
+versions=("$1" "$2")
+files=(
+  "pkg/features/kube_features.go"
+  "staging/src/k8s.io/apiserver/pkg/features/kube_features.go"
+  "staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go"
+  "staging/src/k8s.io/controller-manager/pkg/features/kube_features.go"
+)
+
+out_dir=dev/temp
+mkdir -p "${out_dir}"
+
+for version in "${versions[@]}"; do
+  rm -f "${out_dir}/featuregates-${version}.txt"
+  touch "${out_dir}/featuregates-${version}.txt"
+
+  for file in "${files[@]}"; do
+    { wget -q -O - "https://raw.githubusercontent.com/kubernetes/kubernetes/release-${version}/${file}" || echo; } > "${out_dir}/kube_features.go"
+    grep -E '{Default: .*, PreRelease: .*},' "${out_dir}/kube_features.go" | awk '{print $1}' | { grep -Eo '[A-Z]\w+' || true; } > "${out_dir}/constants.txt"
```

Comment:
I think, we could add some special handling for `LockToDefault` to make the validation even more solid.
E.g. `EndpointSlice` is GA and locked to `true`, so setting `--feature-gates EndpointSlice=false` will make the k8s components fail:
```
$ docker run k8s.gcr.io/kube-apiserver:v1.21.1 /usr/local/bin/kube-apiserver --feature-gates EndpointSlice=false
[...]
Error: invalid argument "EndpointSlice=false" for "--feature-gates" flag: cannot set feature gate EndpointSlice to false, feature is locked to true
2021/06/04 09:03:17 running command: exit status 1
```

Reply:
We discussed this with @rfranzke and explicitly decided against it. The enhanced "solidity" comes at additional complexity and maintenance cost. Besides min and max versions, it would require maintaining a "locked" version per feature gate and even the value to which it's locked, to be able to perform the check. Then the comparison script would need to extract this information (not very trivial and also perhaps a bit fragile having in mind it only uses regexes), and the maintainer would have to maintain it for every new version. Having in mind that the chance of hitting upon this problem is relatively low, we decided it's not worth it.

Reply:
Ok, I understand. Didn't know, you already discussed this.
Fine for me then.